### PR TITLE
C51-320: Hide delegated activities from "My Activities"

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview

This PR hides the activities from "My Activities" that have been delegated to other contacts.

## Before
pr-before
**Note:** The "Medical Evaluation" activity was assigned to another contact so it should not be visible in "My Activities" tab.

## After
pr-after

### Contact Activity Selector endpoint
In order to hide the activities that have been delegated to another contact we needed to create a new endpoint that would query the activities for the contact and filter out the delegated ones manually. The reason for this is because the query to exclude these activities is quite complex and not supported by the API:

```sql
WHERE contact_id = 123
AND (contact_id IN assignee_contact_id
OR assignee_contact_id IS NULL)
```

Since the `assignee_contact_id` is a foreign field, working with NULL is not supported.

The implementation for the new service is as follows:

```php
<?php

class CRM_Civicase_Activity_ContactActivitiesSelector {

  const API_DEFAULT_LIMIT = 25;

  public function getActivitiesForContact($params) {
    $newParams = $this->getParamsWithoutOffsetsAndLimits($params);

    $this->addAssigneeContactIdToReturnParams($newParams);

    $activities = civicrm_api3('Activity', 'get', $newParams);

    if ($activities['error']) {
      return $activities;
    }

    $this->filterOutActivitiesNotBelongingToContact($activities, $newParams);
    $this->paginateActivityRecords($activities, $params);

    return $activities;
  }

  private function getParamsWithoutOffsetsAndLimits($params) {
    $options = CRM_Utils_Array::value('options', $params, []);
    $options['limit'] = 0;
    $options['offset'] = 0;
    $params['options'] = $options;

    return $params;
  }

  private function addAssigneeContactIdToReturnParams(&$params) {
    $return = (array) CRM_Utils_Array::value('return', $params, []);
    $return[] = 'assignee_contact_id';
    $return = array_unique($return);
    $return = implode(',', $return);
    $params['return'] = $return;
  }

  private function filterOutActivitiesNotBelongingToContact(&$activities, $params) {
    $activities['values'] = array_filter($activities['values'], function ($activity) use ($params) {
      $hasNoAssignee = !isset($activity['assignee_contact_id']) || empty($activity['assignee_contact_id']);
      $isContactAssignedToActivity = in_array($params['contact_id'], $activity['assignee_contact_id']);

      return $hasNoAssignee || $isContactAssignedToActivity;
    });

    $activities['count'] = count($activities['values']);
  }

  private function paginateActivityRecords(&$activities, $params) {
    $options = CRM_Utils_Array::value('options', $params, []);
    $limit = CRM_Utils_Array::value('limit', $options, self::API_DEFAULT_LIMIT);
    $offset = CRM_Utils_Array::value('offset', $options, 0);
    $limit = $limit === 0 ? NULL : $limit;

    $activities['values'] = array_slice($activities['values'], $offset, $limit);
  }
}

```

The flow of the service is as follows:

1. It removes the limits and offsets from the original request parameters.
2. It adds the `assignee_contact_id` field to the return list since this field is important to determine which activities have been delegated.
3. It requests all activities using the original request parameters, but without limits or offsets.
4. It removes activities that have been delegated to other users.
5. It paginates the records manually after they have been fetched.

This service was implemented as an Activity action:

```php
<?php
function civicrm_api3_activity_getcontactactivities ($params) {
  $contactActivitySelector = new CRM_Civicase_Activity_ContactActivitiesSelector();

  return $contactActivitySelector->getActivitiesForContact($params);
}

function _civicrm_api3_activity_getcontactactivities_spec(&$params) {
  _civicrm_api3_activity_get_spec($params);

  $params['case_id']['api.required'] = TRUE;
  $params['contact_id']['api.required'] = TRUE;
  $params['return'] = [
    'api.required' => TRUE,
    'api.default' => "assignee_contact_id",
    'name' => 'return',
    'title' => 'Return fields',
    'description' => 'The "assignee_contact_id" field is required for this action.',
  ];
}
```

The spec "extends" from the original `get` action since it should support the same parameters. In order to limit the amount of activity records to be filtered the case and contact id are required. The return is also required so it includes the `assignee_contact_id` field.

### Front end changes

The front-end changes were fairly simple:

when the `involvingContact` is set to `myActivities` the action is changed to `getcontactactivities`, otherwise it uses the normal `get` action:

```js
function loadActivities () {
  var apiAction = $scope.filters['@involvingContact'] === 'myActivities'
    ? 'getcontactactivities'
    : 'get';

  // ...

  return crmApi({
    acts: ['Activity', apiAction, ...],
    all: ['Activity', apiAction, ...]
  });
}
```

### CODE REVIEW

File: api/v3/Activity/Getcontactactivities.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Defines the fields for the "getcontactactivities" action api.
+ * The case id, contact id, and return fields are set to be required.
+ *
+ * @param array $params
+ */
+function _civicrm_api3_activity_getcontactactivities_spec(&$params) {
+  _civicrm_api3_activity_get_spec($params);
+
+  $params['case_id']['api.required'] = TRUE;
```
Comment (deb1990) [inline api/v3/Activity/Getcontactactivities.php]: It means `case_id` is a required field right? But in the dashboard's activity tab, we show all activities irrespective of the case. Are you sure its working in the dashboard?

File: ang/civicase/ActivityFeed.js [hunk 2]
```diff
@@ -264,6 +264,9 @@
      * @return {Promise}
      */
     function loadActivities () {
+      var apiAction = $scope.filters['@involvingContact'] === 'myActivities'
```
Comment (deb1990) [inline ang/civicase/ActivityFeed.js]: Please add a comment explaining why two different api's are necessary.

File: CRM/Civicase/Activity/ContactActivitiesSelector.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+   * the requested one. It also updates the activities count in order to reflect the
+   * new value.
+   *
+   * @param array $activites
+   * @param array $params
+   * @return array
+   */
+  private function filterOutActivitiesNotBelongingToContact(&$activities, $params) {
+    $activities['values'] = array_filter($activities['values'], function ($activity) use ($params) {
+      $hasNoAssignee = !isset($activity['assignee_contact_id']) || empty($activity['assignee_contact_id']);
```
Comment (ajesamson) [inline CRM/Civicase/Activity/ContactActivitiesSelector.php]: I think `empty($activity['assignee_contact_id'])` should be sufficient here 🤷‍♂️ . 
You can confirm if [these](https://secure.php.net/manual/en/function.empty.php#refsect1-function.empty-returnvalues) satisfies your requirement.

File: CRM/Civicase/Activity/ContactActivitiesSelector.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      return $hasNoAssignee || $isContactAssignedToActivity;
+    });
+
+    $activities['count'] = count($activities['values']);
+  }
+
+  /**
+   * Paginates the activity records according to the limit and offset params.
+   *
+   * @param array $activites
```
Comment (ajesamson) [inline CRM/Civicase/Activity/ContactActivitiesSelector.php]: Typographical error on the `$activites`. Should be `$activities`

File: api/v3/Activity/Getcontactactivities.php [hunk 5]
```diff
@@ -20,10 +20,13 @@ function _civicrm_api3_activity_getcontactactivities_spec(&$params) {
 }
 /**
  * Returns the activities for the given contact, limited to a specific case.
+ * This action is needed because the default "get" action does not filter out
+ * activities that have been deletaged to another contact and can't be queried
```
Comment (ajesamson) [inline api/v3/Activity/Getcontactactivities.php]: Guess you meant `delegated` here and not `deletaged`

**Pull request merged**